std/socket.zig

```diff
@@ -0,0 +1,678 @@
+            Protocol.IPV6 => if (is_posix) system.PROTO_ipv6 else system.IPPROTO_IPV6,
+            Protocol.RAW => if (is_posix) system.PROTO_raw else system.IPPROTO_RAW,
+            Protocol.ICMP => if (is_posix) system.PROTO_icmp else system.IPPROTO_ICMP,
+        });
+    }
+};
+
+pub const Shutdown = enum {
+    Read = if (is_posix) system.SHUT_RD else system.SD_RECEIVE,
+    Write = if (is_posix) system.SHUT_WR else system.SD_SEND,
+    Both = if (is_posix) system.SHUT_RDWR else system.SD_BOTH,
+
+    fn toInt(self: Shutdown, comptime T: type) T {
+        return @intCast(T, switch (self) {
+            Shutdown.Read => if (is_posix) system.SHUT_RD else system.SD_RECEIVE,
+            Shutdown.Write => if (is_posix) system.SHUT_WR else system.SD_SEND,
+            Shutdown.Both => if (is_posix) system.SHUT_RDWR else system.SD_BOTH,
+        });
+    }
+};
+
+// pub const Level = enum(u32) {
+//     IP = if (is_posix) system.SOL_IP ,
+//     IPV6 = if (is_posix) system.SOL_IPV6 ,
+//     RM = if (is_posix) system. ,
+//     TCP = if (is_posix) system. ,
+//     UDP = if (is_posix) system. ,
+//     IPX = if (is_posix) system. ,
+//     AppleTalk = if (is_posix) system. ,
+//     IRLMP = if (is_posix) system. ,
+//     Socket = if (is_posix) system. ,
+// };
+
+fn accept4Windows(fd: SocketFd, addr: ?*net.OsAddress, addrlen: ?*system.socklen_t, flags: u32) SocketFd {
+    const result = system.accept(fd, addr, @ptrCast(*c_int, addrlen));
+    const ioctl_mode = @boolToInt((@intCast(c_ulong, flags) & system.FIONBIO) == system.FIONBIO);
+    const no_inherit = (flags & system.WSA_FLAG_NO_HANDLE_INHERIT);
+
+    if (system.ioctlsocket(result, system.FIONBIO, &@intCast(c_ulong, ioctl_mode)) != 0) {
+        return InvalidSocketFd;
+    }
+
+    if (no_inherit == system.WSA_FLAG_NO_HANDLE_INHERIT) {
+        const handle = @intToPtr(system.HANDLE, result);
+        if (system.SetHandleInformation(handle, system.HANDLE_FLAG_INHERIT, 0) == system.FALSE) {
+            return InvalidSocketFd;
+        }
+    }
+
+    return result;
+}
+
+// add timeouts
+
+pub const Socket = struct {
+    fd: SocketFd,
+    domain: Domain,
+    socket_type: SocketType,
+    protocol: Protocol,
+
+    fn initSocket(fd: SocketFd, domain: Domain, socket_type: SocketType, protocol: Protocol) Socket {
+        return Socket{
+            .fd = @intCast(SocketFd, rc),
+            .domain = domain,
+            .socket_type = socket_type,
+            .protocol = protocol,
+        };
+    }
+
+    pub fn new(domain: Domain, socket_type: SocketType, protocol: Protocol) SocketError!Socket {
+        if (is_posix) {
+            const rc = system.socket(domain.toInt(u32), socket_type.toInt(u32), try protocol.toInt(u32));
+            const err = system.getErrno(rc);
+            switch (err) {
+                0 => return initSocket(@intCast(SocketFd, rc), domain, socket_type, protocol),
+                system.EACCES => return SocketError.PermissionDenied,
```

Comment:
Should be able to write just `.EACCES` here.

Reply:
`system.EACCES` is a constant not an enum.